gateway/ln-gateway/src/lib.rs

```diff
@@ -638,6 +651,76 @@ impl Gateway {
         unimplemented!("Restore is not currently supported");
     }
 
+    pub async fn handle_set_configuration_msg(
+        &self,
+        SetConfigurationPayload { password }: SetConfigurationPayload,
+    ) -> Result<()> {
+        let gw_state = self.state.read().await.clone();
+        let can_set_config = matches!(gw_state, GatewayState::Running { .. });
+        if !can_set_config {
+            return Err(GatewayError::Disconnected);
+        }
+
+        let gateway_config = GatewayConfiguration { password };
+
+        let mut dbtx = self.gateway_db.begin_transaction().await;
+        dbtx.insert_entry(&GatewayConfigurationKey, &gateway_config)
+            .await;
+        dbtx.commit_tx().await;
+        info!("Overwrote GatewayConfiguration successfully");
+
+        Ok(())
+    }
+
+    /// This function will return a `GatewayConfiguration` one of three
+    /// different ways. To avoid conflicting configs, the below order is the
+    /// order in which the gateway will respect configurations:
+    /// - `GatewayConfiguration` is read from the database.
+    /// - All cli or environment variables are set such that we can create a
+    ///   `GatewayConfiguration`
+    /// - The `set_configuration` API endpoint writes a `GatewayConfiguration`
+    ///   into the database.
+    pub async fn get_gateway_configuration(&mut self) -> Result<GatewayConfiguration> {
+        {
+            let mut dbtx = self.gateway_db.begin_transaction().await;
+            if let Some(gateway_config) = dbtx.get_value(&GatewayConfigurationKey).await {
+                info!("Gateway using configuration from database.");
+                return Ok(gateway_config);
+            }
+
+            if let Some(password) = self.cli_password.clone() {
+                info!("Gateway using configuration from environment.");
+                return Ok(GatewayConfiguration { password });
+            }
+        }
+
+        self.set_gateway_state(GatewayState::Configuring).await;
+
+        let mut task_group = TaskGroup::new();
+        let mut rx = run_config_webserver(self.listen, &mut task_group).await;
+
+        // Wait for the `/set_configuration` API to send a `GatewayConfiguration`
+        info!("Waiting for GatewayConfiguration to be set");
+        match rx.recv().await {
```

Comment:
Do we actually need that receiver or could we [wait for the DB key](https://docs.fedimint.org/fedimint_core/db/struct.Database.html#method.wait_key_exists) and do the insert in the API handler?

Reply:
Forgot about that, I'll use that instead